### PR TITLE
add some instrumentation for collection locking

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

* Added metrics for collection locks:
  - `arangodb_collection_lock_timeouts_exclusive`: Number of lock timeouts when trying to acquire collection exclusive locks 
  - `arangodb_collection_lock_timeouts_write`: Number of lock timeouts when trying to acquire collection write locks 
  - `arangodb_collection_lock_acquisition_micros`: Total amount of collection lock acquisition time [μs]
  - `arangodb_collection_lock_acquisition_time`: Total collection lock acquisition time histogram [s]

* Reduce lock timeout on followers to 15 seconds.
  Rationale: we should not have any locking conflicts on followers, generally. Any shard locking should be performed on leaders first, which will then, eventually replicate changes to followers. replication to followers is only done once the locks have been acquired on the leader(s). 

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backports required for: 3.7

### Testing & Verification

- [x] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12643/

### CODE REVIEW

**Review by neunhoef — APPROVED** [review]
LGTM

**Review by mchacki — APPROVED** [review]
LGTM

File: arangod/Aql/AqlFunctionFeature.cpp [hunk 1]
```diff
@@ -113,6 +113,11 @@ void AqlFunctionFeature::toVelocyPack(VPackBuilder& builder) {
     } else {
       builder.add(VPackValue("cxx"));
     }
+    // a "stub" here is just a placeholder for a function with no implementation.
+    // this is used for functions which are translated to other constructs during parsing,
+    // e.g. WITHIN_RECTANGLE is translated into a FOR ... FILTER ... loop internally,
+    // so that WITHIN_RECTANGLE is actually never called
+    builder.add("stub", VPackValue(it.second.implementation == &Functions::NotImplemented));
```
Comment (jsteemann) [inline arangod/Aql/AqlFunctionFeature.cpp:120]: removing unrelated change here
```suggestion
```

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp [hunk 2]
```diff
@@ -879,13 +884,15 @@ int RocksDBMetaCollection::doLock(double timeout, AccessMode::Type mode) {
       TRI_ASSERT(false);
       return TRI_ERROR_INTERNAL;
     }
-    if (gotLock) {
-      // keep lock and exit loop
+
+    if (gotLock && waitTime == 0) {
+      // got lock on first iteration!
+      TRI_ASSERT(startTime == 0.0);
+      // keep the lock and exit the loop
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBMetaCollection.cpp:891]:
```suggestion
    if (gotLock) {
      // keep the lock and exit the loop
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/12650/

**Comment by jsteemann** [comment]
Tests blue
